### ISSUE TITLE
subsys/mgmt/hawkbit: Falsely determine that an update is installed successfully

### ISSUE THREAD

**Issue by ycsin** [zephyrproject-rtos/zephyr#37272]
**Describe the bug**
`hawkbit` determines that an update is installed and ready for reboot even though it only downloaded and written 5.75% of the image into the flash.

**To Reproduce**
I testing the `hawkbit` sample with a STM32G0B1RE custom board with a GSM (LTE) modem (Quectel EC21). `image-1` slot is in an external `spi-nor` flash.

**Expected behavior**
Downloads the OTA target firmware successfully, and can be applied to primary slot after `kernel reboot cold`

**Impact**
Unable to use `hawkbit`

**Logs and console output**
[hawkbit_log.txt](https://github.com/zephyrproject-rtos/zephyr/files/6891871/hawkbit_log.txt)

**Environment (please complete the following information):**
 - OS: Windows 10
 - Toolchain: gnuarmemb
 - v2.6.0-rc2-107-gebe282b02d9f

**Additional context**
Ubuntu 20.04, running `hawkbit` in a docker following the `hawkbit`'s sample `README`
`hawkbit` external port is changed to `9010` instead of the default `8080`, sitting behind nginx reverse proxy such that it can be access from hawkbit.mydomain.com with SSL. (shouldn't matter since the device access `hawkbit` using raw IP)

**Comment by ycsin** [comment]
@Navin-Sankar any idea on this?

**Comment by ycsin** [comment]
I don't have a full picture of how this works yet.

I know that it should stay in the `response_cb` and will not reach the [line](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/hawkbit.c#L1211) where it determines whether the download passed or failed. Apparently this check has failed its purpose, so maybe we need something more robust? e.g. Check if the downloaded & written image size matches its actual size?

**Comment by ycsin** [comment]
I notice that `&hb_context.semaphore` isn't used anywhere within the `hawkbit` subsys.

@Navin-Sankar Is that meant to signal the completion of the download in `hawkbit_probe` ?

**Comment by ycsin** [comment]
I changed [L1210](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/hawkbit/hawkbit.c#L1210) from
```c
	if (boot_request_upgrade(BOOT_UPGRADE_TEST)) {
		LOG_ERR("Download failed");
		hb_context.code_status = HAWKBIT_DOWNLOAD_ERROR;
	} else {
		hb_context.code_status = HAWKBIT_UPDATE_INSTALLED;
		hawkbit_device_acid_update(hb_context.json_action_id);
	}
```

to

```c
	if ((boot_request_upgrade(BOOT_UPGRADE_TEST)) ||
	  (k_sem_take(&hb_context.semaphore, K_NO_WAIT) != 0)) {
		LOG_ERR("Download failed");
		hb_context.code_status = HAWKBIT_DOWNLOAD_ERROR;
	} else {
		hb_context.code_status = HAWKBIT_UPDATE_INSTALLED;
		hawkbit_device_acid_update(hb_context.json_action_id);
	}
```

and now the `hawkbit_probe` is able to detect that the download failed and will initiate a new download during the next probe.

**Comment by ycsin** [comment]
Fixed by #37351
